### DISCUSSION TITLE
Memoization with an argument not behaving as I'd expect

### DISCUSSION THREAD

**Discussion by pabuisson** [luckyframework/lucky#1288] [Q&A] [answered]
Hello everyone 👋 Crystal and Lucky newbie here! Having fun with Lucky so far, and one cool thing I noticed in the documentation is the built-in way of memoizing methods, which even takes the method params into account. According to the [documentation](https://luckyframework.org/guides/http-and-routing/routing-and-params#memoization):

```crystal
  # This is only called once per `id` passed in.
  memoize def fetch_user(id : String) : User
    make_api_call && UserQuery.new.find(id)
  end
```

So I tried a dummy example, like this:

```crystal
class Api::Distances::Index < BrowserAction
  route do
    ["one", "one", "one", "one"].each { |value| memoized_method(value) }
    puts "-"*50
    ["two", "three", "two", "three"].each { |value| memoized_method(value) }

    plain_text "Done"
  end

  memoize def memoized_method(key : String) : String
    # something expensive
    puts "memoized method called with #{key}"
    # return some string value to satisfy the method signature
  end
end
```

In my example above, I'd expect to see "one", "two" and "three" written once, the first time the `memoized_method` is called with this param. But instead I get this 🤔 

```
web          | memoized method called with one 
web          | --------------------------------------------------
web          | memoized method called with two
web          | memoized method called with three
web          | memoized method called with two
web          | memoized method called with three
```

The memoization works well as long as I call the memoized method with the same argument, again and again, the method is only run once. BUT calling the method again with a different argument seems to "reset" the memoization? 

Did I misunderstand the documentation, or am I simply making some kind of Crystal-newbie mistake here? 
Sorry for the long post, and thanks for anyone that'll help me decipher this 🙏

**Comment by matthewmcgarvey** [comment, marked as answer]
Hey @pabuisson! It only memoizes the last argument it was called with. Looking at the docs now, I can see how you'd get that impression, though.

**Reply by pabuisson** [reply]
Oh that's what I thought when experimenting, but indeed, given the example in the documentation, I was very confused by the way it behaves. So there's actually no problem here, just that I didn't understand how it behave just reading the doc. Thanks a lot @matthewmcgarvey for answering and giving me a crystal-clear (ha ha!) confirmation about this 👍

**Comment by stephendolan** [comment]
@matthewmcgarvey If you've got a way in mind to make the documentation more clear, would you mind opening up a website issue? Happy to get to that at some point to make things clearer.
